src/meta/src/barrier/rpc.rs

```diff
@@ -28,141 +30,190 @@ use risingwave_common::util::tracing::TracingContext;
-        let stream_rpc_manager = self.stream_rpc_manager.clone();
-        // todo: the collect handler should be abort when recovery.
-        let _join_handle = tokio::spawn(async move {
-            let span = command_context.span.clone();
-            if let Some(prev_inject_rx) = prev_inject_rx {
-                if prev_inject_rx.await.is_err() {
-                    let _ = tx.send(BarrierCollectResult {
-                        prev_epoch,
-                        result: Err(anyhow!("prev barrier failed to be injected").into()),
-                    });
-                    return;
-                }
-            }
-            let result = stream_rpc_manager
-                .inject_barrier(command_context.clone())
-                .instrument(span.clone())
-                .await;
-            match result {
-                Ok(node_need_collect) => {
-                    if let Some(inject_tx) = inject_tx {
-                        let _ = inject_tx.send(());
+    pub(super) fn next_response(
+        &mut self,
+    ) -> impl Future<Output = MetaResult<(WorkerId, u64, StreamingControlStreamResponse)>> + '_
+    {
+        poll_fn(|cx| match self.response_streams.poll_next_unpin(cx) {
+            Poll::Ready(None) | Poll::Pending => Poll::Pending,
+            Poll::Ready(Some((worker_id, response_stream, result))) => match result {
+                Ok(resp) => match &resp.response {
+                    Some(streaming_control_stream_response::Response::CompleteBarrier(_)) => {
+                        self.response_streams
+                            .push(into_future(worker_id, response_stream));
+                        let node = self
+                            .nodes
+                            .get_mut(&worker_id)
+                            .expect("should exist when get collect resp");
+                        let command = node
+                            .inflight_barriers
+                            .pop_front()
+                            .expect("should exist when get collect resp");
+                        Poll::Ready(Ok((worker_id, command.prev_epoch.value().0, resp)))
+                    }
+                    resp => Poll::Ready(Err(anyhow!("get unexpected resp: {:?}", resp).into())),
+                },
+                Err(err) => {
+                    let mut node = self
+                        .nodes
+                        .remove(&worker_id)
+                        .expect("should exist when get collect resp");
+                    warn!(node = ?node.worker, err = ?err.as_report(), "get error from response stream");
+                    if let Some(command) = node.inflight_barriers.pop_front() {
+                        self.context.report_collect_failure(&command, &err);
+                        Poll::Ready(Err(err))
+                    } else {
+                        // for node with no inflight barrier, simply ignore the error
+                        Poll::Pending
```

Comment:
In this case we're not registering a waker but return `Pending`. I'm not sure whether this would be problematic.

Reply:
Is it necessary to write poll function ourselves? Personally I would strongly suggest using `async fn` manner as much as possible.